### PR TITLE
refactor: general cleaning up of the tree module

### PR DESCRIPTION
**Pull request by thatlittleboy**
## Overview

Description of the changes proposed in this pull request:

There shouldn't be any functional changes in this PR, just cleaning up of the code and documentation in 2 files

* `shap/explainers/_tree.py`
* `tests/explainers/test_tree.py` : refactor sklearn into its own class, similar to what was done for LGBM in an earlier PR. EDIT: the purpose of this refactor is purely organizational, please read the discussion [here](https://github.com/dsgibbons/shap/pull/88#discussion_r1210869174)

Mini port of the (now stale) PR: #1306 

## Checklist

- [X] All [pre-commit checks](https://pre-commit.com/#install) pass.
- [x] Unit tests added (if fixing a bug or adding a new feature)
- [ ] Added entry to `CHANGELOG.md` (if changes will affect users)

### CODE REVIEW

File: tests/explainers/test_tree.py
Comment (thatlittleboy) [inline tests/explainers/test_tree.py:24]: Added this test for coverage purposes

File: tests/explainers/test_tree.py
Comment (thatlittleboy) [inline tests/explainers/test_tree.py:962]: And this test also for coverage

File: shap/explainers/_tree.py
Comment (thatlittleboy) [inline shap/explainers/_tree.py]: Corrected this independent -> interventional

File: shap/explainers/_tree.py
Comment (thatlittleboy) [inline shap/explainers/_tree.py:79]: Generally added code formatting in the docstrings, and some rephrasing where necessary.

**Review by thatlittleboy — COMMENTED** [review]
Sorry for the large PR. I tried to highlight the key changes, the rest should just be formatting changes.

File: tests/explainers/test_tree.py
Comment (connortann) [inline tests/explainers/test_tree.py:755]: Out of interest, what is the motivation behind grouping these tests into a class? Is it purely organisational, or are there other benefits?

File: shap/explainers/_tree.py
Comment (connortann) [inline shap/explainers/_tree.py:367]: What is the purpose of this refactor? IMO the original is clearer: directly putting the string in the call to `ValueError`.

I seem to remember that there exists a linter rule that prohibits exception messages over a certain length. However, I don't think that is relevant in our code base as the relevant linting rule is not enabled (and I would not want to enable it).

**Comment by connortann** [comment]
Thanks for the PR! It look likes a good cleanup.

There is a mix of useful functional changes (e.g. moving tests to classes, adding tests, changing documentation), and purely stylistic changes (e.g. adding spacing, commas, newlines).  Stylistic changes are largely down to personal preference and prone to "bike-shedding", and lead to a large diff.

I would like to propose the general rule:

1. PRs with non-style changes should generally avoid any style changes to existing code, and
2. PRs with style changes should have _zero_ non-style changes, and ideally should be implemented with automated tools to remove risk of human error (e.g. ruff, isort, black, etc).

I'd be happy to approve this PR once other threads are resolved. However, for future PRs I'd like to suggest we adopt a more strict separation of style & non-style changes.

File: shap/explainers/_tree.py
Comment (connortann) [reply shap/explainers/_tree.py:367]: (Resolved in other thread)

File: tests/explainers/test_tree.py
Comment (dsgibbons) [reply tests/explainers/test_tree.py:755]: @connortann we have discussed something similar before. See https://github.com/dsgibbons/shap/pull/88#discussion_r1210869174.

**Comment by thatlittleboy** [comment]
Thanks @connortann , noted on the comment regarding style vs functional. Will come back to this PR and address the other comments later this week.

File: tests/explainers/test_tree.py
Comment (connortann) [reply tests/explainers/test_tree.py:755]: Thanks for sharing the link! I agree with your comments in that thread.

FWIW, I'd be hestitant to make the tests themselves object-oriented (e.g. with class-based setup and teardown), as that is contrary to the functional pytest approach. But it seems that's not the intent here, so it looks good.

Comment (connortann) [reply tests/explainers/test_tree.py:755]: PS - one rather minor downside of adding the class-based approach is a complex git history. As with all style changes that cause a big diff, it will be harder to see the evolution of particular tests with `git blame`.

One mitigation could be to add a `.git-blame-ignore-revs` file with a list of commits to ignore. This is another reason to aim for a strict separation style PRs from PRs with functional changes. For more info see:
- python black: [Avoiding ruining git blame](https://black.readthedocs.io/en/stable/guides/introducing_black_to_your_project.html#avoiding-ruining-git-blame)
- Github: [ignore commits in blame view](https://docs.github.com/en/repositories/working-with-files/using-files/viewing-a-file#ignore-commits-in-the-blame-view)

**Review by connortann — APPROVED** [review]
LGTM! I'll leave it to you to merge when you're ready.

**Pull request merged**
